android/app/src/main/java/com/microsoft/codepush/react/CodePush.java

```diff
@@ -439,19 +441,43 @@ public void run() {
         }
 
         @ReactMethod
-        public void downloadUpdate(final ReadableMap updatePackage, final Promise promise) {
+        public void downloadUpdate(final ReadableMap updatePackage, final boolean notifyProgress, final Promise promise) {
             AsyncTask<Void, Void, Void> asyncTask = new AsyncTask<Void, Void, Void>() {
                 @Override
                 protected Void doInBackground(Void... params) {
                     try {
                         WritableMap mutableUpdatePackage = CodePushUtils.convertReadableMapToWritableMap(updatePackage);
                         mutableUpdatePackage.putString(BINARY_MODIFIED_TIME_KEY, "" + getBinaryResourcesModifiedTime());
                         codePushPackage.downloadPackage(mutableUpdatePackage, CodePush.this.assetsBundleFileName, new DownloadProgressCallback() {
+                            private boolean nextFrameBusy = false;
+                            private DownloadProgress latestDownloadProgress = null;
+
                             @Override
                             public void call(DownloadProgress downloadProgress) {
-                                getReactApplicationContext()
-                                        .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
-                                        .emit(DOWNLOAD_PROGRESS_EVENT_NAME, downloadProgress.createWritableMap());
+                                if (!notifyProgress) {
+                                    return;
+                                }
+
+                                this.latestDownloadProgress = downloadProgress;
+                                if (nextFrameBusy) {
+                                    return;
+                                }
+
+                                nextFrameBusy = true;
+                                mainActivity.runOnUiThread(new Runnable() {
```

Comment:
postFrameCallback needs to run on the UI thread, otherwise it throws this error "Thread does not have a looper"